File: Sources/UpdateChecker.swift
import AppKit
import DictCore
import Foundation

/// Checks dict.tianxu.cloud/version for a newer version.
/// If the remote version differs, prompts the user to download.
enum UpdateChecker {
    private static let currentVersion: String = {
        if let v = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String {
            return v
        }
        // Fallback: read from Info.plist on disk (when running outside .app bundle)
        let plistPath = Bundle.main.bundlePath + "/../Resources/Info.plist"
        if let dict = NSDictionary(contentsOfFile: plistPath),
           let v = dict["CFBundleShortVersionString"] as? String {
            return v
        }
        return "0.0.0"
    }()
    private static let versionURL = URL(string: "https://dict.tianxu.cloud/version")!
    private static let downloadURL = URL(string: "https://dict.studio/Dict.dmg")!

    /// - Parameter silent: When true, only shows alert if update is available (used on launch).
    ///                     When false, always shows feedback (used for manual check).
    static func checkForUpdate(silent: Bool = false) {
        var request = URLRequest(url: versionURL)
        request.timeoutInterval = 10

        URLSession.shared.dataTask(with: request) { data, _, error in
            DispatchQueue.main.async {
                if let error = error {
                    Log.info("Update check failed: \(error.localizedDescription)")
                    if !silent {
                        showErrorAlert()
                    }
                    return
                }

                guard let data = data,
                      let remoteVersion = String(data: data, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines),
                      !remoteVersion.isEmpty
                else {
                    if !silent { showErrorAlert() }
                    return
                }

                if isNewerVersion(remoteVersion, than: currentVersion) {
                    Log.info("Update available: v\(remoteVersion) (current: v\(currentVersion))")
                    showUpdateAlert(version: remoteVersion)
                } else {
                    Log.info("Up to date (v\(currentVersion))")
                    if !silent {
                        showUpToDateAlert()
                    }
                }
            }
        }.resume()
    }

    /// Returns true if `remote` is a higher semver than `current`.
    private static func isNewerVersion(_ remote: String, than current: String) -> Bool {
        DictCore.isNewerVersion(remote, than: current)
    }

    private static func showUpdateAlert(version: String) {
        let alert = NSAlert()
        alert.messageText = "Dict v\(version) Available"
        alert.informativeText = "A new version of Dict is available. You're running v\(currentVersion)."
        alert.addButton(withTitle: "Download")
        alert.addButton(withTitle: "Later")
        alert.alertStyle = .informational

        if alert.runModal() == .alertFirstButtonReturn {
            NSWorkspace.shared.open(downloadURL)
        }
    }

    private static func showUpToDateAlert() {
        let alert = NSAlert()
        alert.messageText = "You're Up to Date"
        alert.informativeText = "Dict v\(currentVersion) is the latest version."
        alert.addButton(withTitle: "OK")
        alert.alertStyle = .informational
        alert.runModal()
    }

    private static func showErrorAlert() {
        let alert = NSAlert()
        alert.messageText = "Update Check Failed"
        alert.informativeText = "Could not connect to the update server. Please try again later."
        alert.addButton(withTitle: "OK")
        alert.alertStyle = .warning
        alert.runModal()
    }
}

File: src-tauri/src/update_checker.rs
use crate::settings::is_newer_version;
use reqwest::Client;
use std::time::Duration;

const VERSION_URL: &str = "https://dict.tianxu.cloud/version";
const DOWNLOAD_URL: &str = "https://dict.studio/Dict.dmg";
const CURRENT_VERSION: &str = env!("CARGO_PKG_VERSION");

#[derive(Debug, Clone, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub enum UpdateResult {
    UpdateAvailable {
        remote_version: String,
        current_version: String,
        download_url: String,
    },
    UpToDate {
        current_version: String,
    },
    Error {
        message: String,
    },
}

/// Check for updates by comparing local version against remote
pub async fn check_for_update() -> UpdateResult {
    let client = Client::new();

    let resp = match client
        .get(VERSION_URL)
        .timeout(Duration::from_secs(10))
        .send()
        .await
    {
        Ok(r) => r,
        Err(e) => {
            tracing::warn!("Update check failed: {}", e);
            return UpdateResult::Error {
                message: "Could not connect to the update server.".to_string(),
            };
        }
    };

    let remote_version = match resp.text().await {
        Ok(t) => t.trim().to_string(),
        Err(e) => {
            tracing::warn!("Failed to read version response: {}", e);
            return UpdateResult::Error {
                message: "Could not read version response.".to_string(),
            };
        }
    };

    if remote_version.is_empty() {
        return UpdateResult::Error {
            message: "Empty version response.".to_string(),
        };
    }

    if is_newer_version(&remote_version, CURRENT_VERSION) {
        tracing::info!(
            "Update available: v{} (current: v{})",
            remote_version,
            CURRENT_VERSION
        );
        UpdateResult::UpdateAvailable {
            remote_version,
            current_version: CURRENT_VERSION.to_string(),
            download_url: DOWNLOAD_URL.to_string(),
        }
    } else {
        tracing::info!("Up to date (v{})", CURRENT_VERSION);
        UpdateResult::UpToDate {
            current_version: CURRENT_VERSION.to_string(),
        }
    }
}
